### PR TITLE
Core/Spells: implement Mad/Crazy Alchemist's potions

### PR DESCRIPTION
**Pull request by Gerhood**
**Changes proposed:**

Make Mad and Crazy Alchemist's potion work. For this we need to fix triggered potion spells first.

**Target branch(es):** 3.3.5

**Tests performed:** The potions are working fully blizzlike, info was taken from wowhead comments.

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -21725,7 +21725,12 @@ void Player::UpdatePotionCooldown(Spell* spell)
     }
     // from spell cases (m_lastPotionId set in Spell::SendSpellCooldown)
     else
-        GetSpellHistory()->SendCooldownEvent(spell->m_spellInfo, m_lastPotionId, spell);
+    {
```
Comment (ariel-) [inline src/server/game/Entities/Player/Player.cpp]: Wait, what is this unrelated change?

File: src/server/game/Spells/SpellEffects.cpp [hunk 2]
```diff
@@ -1742,6 +1742,86 @@ void Spell::EffectEnergize(SpellEffIndex effIndex)
...
+
+        bool useElixir = true;
+
+        SpellGroup chosenSpellGroup = SPELL_GROUP_NONE;
+        if (sSpellMgr->IsSpellMemberOfSpellGroup(chosenElixir, SPELL_GROUP_ELIXIR_BATTLE))
+            chosenSpellGroup = SPELL_GROUP_ELIXIR_BATTLE;
+        if (sSpellMgr->IsSpellMemberOfSpellGroup(chosenElixir, SPELL_GROUP_ELIXIR_GUARDIAN))
+            chosenSpellGroup = SPELL_GROUP_ELIXIR_GUARDIAN;
+        // If another spell of the same group is already active the elixir should not be cast
+        if (chosenSpellGroup) {
```
Comment (robinsch) [inline src/server/game/Spells/SpellEffects.cpp]: style

File: src/server/game/Spells/SpellEffects.cpp [hunk 3]
```diff
@@ -1742,6 +1742,86 @@ void Spell::EffectEnergize(SpellEffIndex effIndex)
     if (!unitTarget->IsAlive())
         return;
 
+    // Handle Secondary Potion Effects here
+    // Mad Alchemist's Potion (34440)
+    if (m_spellInfo->Id == 45051)
+    {
+        std::set<uint32> availableElixirs;
```
Comment (ariel-) [inline src/server/game/Spells/SpellEffects.cpp]: This should really be in a spell script.
Also, default container for TC is std::vector.

Also, you can use initializer-list style in c++11:
```
std::vector<uint32> availableElixirs =
{
    33720, // Onslaught Elixir (28102)
    54452, // Adept's Elixir (28103)
(....)
    28489,
    28496
};

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (Gerhood) [inline src/server/game/Entities/Player/Player.cpp]: We must not set lastpotion for the triggered (second) spell.

File: src/server/game/Spells/SpellEffects.cpp (hunk 3 above)
Comment (Gerhood) [inline src/server/game/Spells/SpellEffects.cpp]: Would you mind telling me wich spell script function to use? I dont know much about spell scripts yet :/

Comment (ariel-) [inline src/server/game/Spells/SpellEffects.cpp]: Yes, there are hooks to which you can attach your functions, for example you can use an AfterCast hook. If in doubt, search for other script uses to see how it works, but you need to put this in your register subroutine:

```c++
void Register() override
{
    AfterCast += SpellCastFn(yourscriptclassname::yourfunction);
}
```

Comment (Gerhood) [inline src/server/game/Spells/SpellEffects.cpp]: Ok, I'll see for it later today or this week

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 4]
```diff
@@ -0,0 +1,4 @@
+DELETE FROM spell_script_names WHERE spell_id IN (45051, 53750);
+INSERT INTO spell_script_names (spell_id, ScriptName) VALUES
+(45051, 'spell_item_mad_alchemists_potion'),
+(53750, 'spell_item_crazy_alchemists_potion');
```
Comment (Aokromes) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: missing newline.

**Comment by Aokromes** [comment]
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
Conflicting files
src/server/scripts/Spells/spell_item.cpp

**Comment by Gerhood** [comment]
Ill do merging when the PR has been accepted

**Comment by tkrokli** [comment]
@Gerhood : that this PR is still open and also labelled with 4 category labels should be enough to tell you that this PR is accepted in its original form, as well as the fact that others have taken their time to comment on the code (either to suggest improvements or ask what a code change will do to the source).

To be fair, the fact that Aokromes asked you to fix the file conflict also shows that this is still a valid PR.
I find it difficult to see what more you need to see that it is accepted, short of merging it (which can't be done until you fix the file conflict and/or get more improvement suggestions from TC developers).

**Comment by Gerhood** [comment]
I understand what you are saying and with accepting I actually meant being ready for merging, so there was a bit of a misunderstanding.
Also Aokromes' unelaborate comment did not make me more confident that PR will be merged anytime soon

**Comment by Aokromes** [comment]
I only pointed the pr had conflicts, i never said it will be not merged.

**Comment by Gerhood** [comment]
That's not what I meant but never mind :)

**Comment by Aokromes** [comment]
@Gerhood can you update again and enable allow contributors to edit pull requests? thx.

**Comment by Gerhood** [comment]
Note on spells with the TRIGGERED_IGNORE_SPELL_AND_CATEGORY_CD flag (IIRC):

**On head:** Spell cooldowns are  _checked_ (!) but _not set_, for potions spells, m_lastPotionId ist _set and checked_

**In this PR:** For spells from potions m_lastPotionId ist _neither checked nor set_
